resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ResourceTypeValidator.scala

```diff
@@ -0,0 +1,162 @@
+    }
+
+    val (standardResourceTypeId: String, customResourceTypeId: String) =
+      getResourceTypeIdsByRole(role, mode, resourceType)
+
+    if (resourceTypeObj != null && customResourceTypes.contains(customResourceTypeId)) {
+      return formatDuplicateResourceTypeErrorMessage(standardResourceTypeId, customResourceTypeId)
+    }
+    ""
+  }
+
+  private def formatDuplicateResourceTypeErrorMessage(standardResourceTypeId: String,
+                                                      customResourceTypeId: String): String = {
+    s"$standardResourceTypeId and $customResourceTypeId" +
+      " configs are both present, only one of them is allowed at the same time!"
+  }
+
+  private def getResourceTypeIdsByRole(role: String, mode: String, resourceType: String) = {
+    val standardResourceTypeId: String = s"spark.$role.$resourceType"
+
+    var customResourceTypeId: String = ""
+    (role, mode) match {
+      case ("driver", "client") => customResourceTypeId += "spark.yarn.am.resource."
+      case ("driver", "cluster") => customResourceTypeId += "spark.yarn.driver.resource."
+      case ("executor", _) => customResourceTypeId += "spark.yarn.executor.resource."
+    }
+
+    customResourceTypeId += resourceType
+
+    (standardResourceTypeId, customResourceTypeId)
+  }
+
+  private class ResourceTypeConfigProperties(val role: String,
+                                             val mode: String = "",
+                                             val resourceType: String)
+
+
+  private class RequestedResources(val sparkConf: SparkConf) {
+    val driverMemory: String = safelyGetFromSparkConf(sparkConf, "spark.driver.memory")
+    val driverCores: String = safelyGetFromSparkConf(sparkConf, "spark.driver.cores")
+    val executorMemory: String = safelyGetFromSparkConf(sparkConf, "spark.executor.memory")
+    val executorCores: String = safelyGetFromSparkConf(sparkConf, "spark.executor.cores")
+    val customResourceTypesForDriverClientMode: Map[String, String] =
+      extractCustomResourceTypes(sparkConf, "spark.yarn.am.resource.")
+    val customResourceTypesForDriverClusterMode: Map[String, String] =
+      extractCustomResourceTypes(sparkConf, "spark.yarn.driver.resource.")
+    val customResourceTypesForExecutor: Map[String, String] =
+      extractCustomResourceTypes(sparkConf, "spark.yarn.executor.resource.")
+
+    private def extractCustomResourceTypes(sparkConf: SparkConf,
+                                           propertyPrefix: String): Map[String, String] = {
+      val result: collection.mutable.HashMap[String, String] =
+        new collection.mutable.HashMap[String, String]()
+
+      val propertiesWithPrefix: Array[(String, String)] = sparkConf.getAllWithPrefix(propertyPrefix)
+      propertiesWithPrefix.foreach(e => result.put(propertyPrefix + e._1, e._2))
+
+      result.toMap
+    }
+
+    private def safelyGetFromSparkConf(sparkConf: SparkConf, key: String): String = {
```

Comment:
This is `conf.getOption(key).orNull`. It also does not check for default values. For example, the driver's memory has a config constant (`DRIVER_MEMORY`) which encapsulates its default value.

Reply:
You mean this one, right? org.apache.spark.internal.config#DRIVER_MEMORY
Since I see no usage of this constant, neither EXECUTOR_MEMORY in code, I’m a bit confused.
Anyway, I modified the code and started to use these constants for accessing the values, as you suggested.